render_hawk_2/renderer_class.cs

```diff
@@ -0,0 +1,466 @@
+            sphere_hit = get_closest_sphere (scene, position, ref vector_s, sphere_hit);
+
+            if (vector_s != vector
+             && (vector_t == vector
+              || vector_class.vector_is_larger (vector_t, vector_s)))
+            {
+                vector = vector_s;
+                triangle_hit = -1;
+
+                if (sphere_hit != -1)
+                {
+                    for (int n = 0; n < 3; n++)
+                    {
+                        total_color_radiance[n] += scene.material[scene.sphere[sphere_hit].material_index - 1].color[n].radiance;
+                    }
+
+                    if (i != scene.ray_depth - 1)
+                    {
+                        position = vector_class.vector_add (position, vector_class.vector_scale (vector, 1 - 0.000000001));
+                        vector = scene.material[scene.sphere[sphere_hit].material_index - 1].get_next_direction_sphere (position, vector, scene.sphere[sphere_hit], ref medium_refractive_index, ref multiplier);
+                    }
+                }
+                else
+                {
+                    for (int n = 0; n < 3; n++)
+                    {
+                        total_color_radiance[n] += scene.sky_radiance[n];
+                    }
+
+                    break;
+                }
+            }
+            else
+            {
+                vector = vector_t;
+                sphere_hit = -1;
+
+                if (triangle_hit != -1)
+                {
+                    for (int n = 0; n < 3; n++)
+                    {
+                        total_color_radiance[n] += scene.material[scene.triangle[triangle_hit].material_index - 1].color[n].radiance;
+                    }
+
+                    if (i != scene.ray_depth - 1)
+                    {
+                        position = vector_class.vector_add (position, vector_class.vector_scale (vector, 1 - 0.000000001));
+                        vector = scene.material[scene.triangle[triangle_hit].material_index - 1].get_next_direction_triangle (vector, scene.triangle[triangle_hit], ref medium_refractive_index, ref multiplier);
+                    }
+                }
+                else
+                {
+                    for (int n = 0; n < 3; n++)
+                    {
+                        total_color_radiance[n] += scene.sky_radiance[n];
+                    }
+
+                    break;
+                }
+            }
+        }
+
+
+        for (int i = 0; i < 3; i++)
+        {
+            total_color_radiance[i] *= multiplier[i];
+        }
+
+        return total_color_radiance;
+    }
+
+    private static int get_closest_sphere (scene_class scene, double[] position, ref double[] vector, int exclude)
+    {
+        int output_sphere = -1;
+        double t_previous = double.MaxValue;
+        double t = 0;
+
+        for (int i = 0; i < scene.sphere.Count; i++)
+        {
+            if (i != exclude)

```

Comment:
To handle self-illumination for potential future concave primitives, an epsilon value would be necessary. I wonder how other rendering engines handle this kind of problem.